sky/data/storage_utils.py

```diff
@@ -120,91 +123,99 @@ def get_excluded_files_from_gitignore(src_dir_path: str) -> List[str]:
     """
     expand_src_dir_path = os.path.expanduser(src_dir_path)
 
-    git_exclude_path = os.path.join(expand_src_dir_path, '.git/info/exclude')
-    gitignore_path = os.path.join(expand_src_dir_path,
-                                  constants.GIT_IGNORE_FILE)
+    # We will use `git ls-files` to list files that we should ignore, but
+    # `ls-files` will not recurse into subdirectories. So, we need to manually
+    # list the submodules and run `ls-files` within the root and each submodule.
+    # Print the submodule paths relative to expand_src_dir_path, separated by
+    # null chars.
+    submodules_cmd = (f'git -C {shlex.quote(expand_src_dir_path)} '
+                      'submodule foreach -q "printf \\$displaypath\\\\\\0"')
+
+    try:
+        submodules_output = subprocess.run(submodules_cmd,
+                                           shell=True,
+                                           stdout=subprocess.PIPE,
+                                           stderr=subprocess.PIPE,
+                                           check=True,
+                                           text=True)
+    except subprocess.CalledProcessError as e:
+        gitignore_path = os.path.join(expand_src_dir_path,
+                                      constants.GIT_IGNORE_FILE)
+
+        if (e.returncode == exceptions.GIT_FATAL_EXIT_CODE and
+                'not a git repository' in e.stderr):
+            # If git failed because we aren't in a git repository, but there is
+            # a .gitignore, warn the user that it will be ignored.
+            if os.path.exists(gitignore_path):
+                logger.warning('Detected a .gitignore file, but '
+                               f'{src_dir_path} is not a git repository. The '
+                               '.gitignore file will be ignored. '
+                               f'{_USE_SKYIGNORE_HINT}')
+            # Otherwise, this is fine and we can exit early.
+            return []
+
+        if e.returncode == exceptions.COMMAND_NOT_FOUND_EXIT_CODE:
+            # Git is not installed. This is fine, skip the check.
+            # If .gitignore is present, warn the user.
+            if os.path.exists(gitignore_path):
+                logger.warning(f'Detected a .gitignore file in {src_dir_path}, '
+                               'but git is not installed. The .gitignore file '
+                               f'will be ignored. {_USE_SKYIGNORE_HINT}')
+            return []
+
+        # Pretty much any other error is unexpected, so re-raise.
+        raise
```

Comment:
Should we just `return []` the error message output with `logger.debug`. Just want to make sure that there is no unexpected case causing `sky jobs launch` to become `FAILED_CONTROLLER`.

Reply:
This should be client-side, so okay to crash if the environment is messed up. It won't cause `FAILED_CONTROLLER`. Double-checked to verify.